### ISSUE TITLE
Getting Integrity errors when multiple users are using the NO system

### ISSUE THREAD

**Issue by acrigney** [NakedObjectsGroup/NakedObjectsFramework#147]
Hi Guys,
               I have upgraded my very old version of our application from 5 years ago to the latest NO9 version and its all working except that when multiple users use the system we get these integrity errors. These errors also used to occur on the old system but the system was only used by 1 person most of  the time before. Now the business wants to have multiple users in multiple offices around Australia using the system.
Here are a few of the errors. 
ERROR 32 NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity failed for: TEOID#10 bad values: UserName:WorkItemId:0CourtType:UnspecifiedCreator:WorkItemState:PendingADUserName:AuthorisingOfficerEmail:IOffice:IsWorkItemApproved:FalseUserName:WorkItemOfficerEmail: old hash:  new hash 8809051539268595353
ERROR 36 NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity failed for: TEOID#11 bad values: UserName:WorkItemId:0CourtType:UnspecifiedCreator:WorkItemState:PendingADUserName:AuthorisingOfficerEmail:IOffice:IsWorkItemApproved:FalseUserName:WorkItemOfficerEmail: old hash:  new hash 8809051539268595353
ERROR 37 NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity failed for: TEOID#12 bad values: UserName:WorkItemId:0CourtType:UnspecifiedCreator:WorkItemState:PendingADUserName:AuthorisingOfficerEmail:IOffice:IsWorkItemApproved:FalseUserName:WorkItemOfficerEmail: old hash:  new hash 8809051539268595353
ERROR 21 NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity failed for: TEOID#13 bad values: UserName:WorkItemId:0CourtType:UnspecifiedCreator:WorkItemState:PendingADUserName:AuthorisingOfficerEmail:IOffice:IsWorkItemApproved:FalseUserName:WorkItemOfficerEmail: old hash:  new hash 8809051539268595353

This is the NO code that generates the errors. Although in the new version the user does not see a pop up error, I guess they are configured not to show unless required, I will check the doco.

private void ValidateTransientIntegrity(ObjectContext nakedObject, string digest) {
            string rawValue;
            var transientHash = GetTransientSecurityHash(nakedObject, out rawValue);

            if (transientHash != digest) {
                Log.Error($"Transient Integrity failed for: {nakedObject.Id} bad values: {rawValue} old hash: {digest} new hash {transientHash}");
                throw new BadRequestNOSException("Values provided may not be persisted as an object (ensure any derived properties are annotated NotPersisted");
            }
Each of the model objects derives from a BaseDomainObject shown here, this is used to log all audit all transactions on any object .

Perhaps the problem is that referenced objects that are not changed should be marked as not persisted.
What do you mean by derived properties here? Not in the sense of inheritance?

using System;
using System.ComponentModel;
using NakedObjects;
using System.ComponentModel.DataAnnotations;
using System.Diagnostics;
using Lansw.Uwas.Common.Enums;
using System.Reflection;

// Toplevel domain object

namespace Lansw.Uwas.Model
{
    public abstract class BaseDomainObject
    {

        #region Injected Services

        public IDomainObjectContainer Container { protected get; set; }
        public AuditRecordRepository AuditRecordRepository { set; protected get; }

        #endregion

        #region Life Cycle Methods

        public virtual void Created()
        {
            // NB no point in logging this as this is called as soon as
            // an object is initialized in the UI!!
            //if (IsAuditableRecord)
            //{
            //    CreateAuditRecord();
            //}
        }

        public virtual void Persisting()
        {
            rowguid = Guid.NewGuid();

            ModifiedDate = DateTime.Now;
            //CreatedDate = DateTime.Now;

            Timestamp = BitConverter.GetBytes(Stopwatch.GetTimestamp());

            // NB no point in creating an audit record here as
            // the ids values will not be set till the object is persisted!
        }

        public virtual void Persisted()
        {
            CreatedDate = DateTime.Now;
            if (IsAuditableRecord)
            {
                CreatePersistAuditRecord();
            }
        }

        public virtual void Updating()
        {
            ModifiedDate = DateTime.Now;
            Timestamp = BitConverter.GetBytes(Stopwatch.GetTimestamp());

            if (IsAuditableRecord)
            {
                CreateUpdateAuditRecord();
            }
        }

        private void WriteAuditRecord(AuditType auditType)
        {
            string notProxyName = null;
            // NB this may be a proxy object may be 

            var auditRecord = AuditRecordRepository.NewAuditRecord();

            auditRecord.AuditType = auditType;

            Type notProxyType = GetNotProxyType(ref notProxyName);

            auditRecord.RecordName = notProxyName;

            if (auditRecord.AddRecordFields(notProxyType, this))
            {
                Container.Persist<AuditRecord>(ref auditRecord);
                Debug.WriteLine(string.Format("Created {0} Audit Record For {1}", auditType, auditRecord.RecordName));
            }
        }

        //[Hidden]
        //public virtual string NotProxyTypeName 
        //{
        //    get
        //    {
        //        string properTypeName;
        //        if (IsProxyEntity())
        //        {
        //            properTypeName = this.GetType().BaseType.Name.Substring(0, this.GetType().Name.IndexOf('_'));
        //        }
        //        else
        //        {
        //            properTypeName = this.GetType().Name;
        //        }
        //        return (properTypeName);
        //    }
        //}

        private Type GetNotProxyType(ref string notProxyName)
        {
            Type type;

            notProxyName = null;
            if (IsProxyEntity())
            {
                type = this.GetType().BaseType;
            }
            else
            {
                type = this.GetType();
            }

            notProxyName = type.Name;
            return (type);
        }

        private void CreateAuditRecord()
        {
            //if (!IsProxyEntity())
            {
                WriteAuditRecord(AuditType.CreateRecord);
            }
        }

        private void CreatePersistAuditRecord()
        {
            // NO passes the proxy object for the persist and update so we can't check that its
            // not a proxy object!!
            //if (!IsProxyEntity())
            {
                WriteAuditRecord(AuditType.PersistRecord);
            }
        }

        private void CreateUpdateAuditRecord()
        {
            // NO passes the proxy object for the persist and update so we can't check that its
            // not a proxy object!!
            //if (!IsProxyEntity())
            {
                WriteAuditRecord(AuditType.UpdateRecord);
            }
        }

        private bool IsProxyEntity()
        {
            return (this.GetType().Namespace == "System.Data.Entity.DynamicProxies");
        }

        #endregion
        [Hidden(WhenTo.Always)]
        [NotPersisted]
        public virtual bool IsAuditableRecord { get { return true; } }

        [Hidden(WhenTo.Always)]
        [NotPersisted]
        public virtual byte[] Timestamp { get; set; }

        [DisplayName("Created Date")]
        [Hidden(WhenTo.UntilPersisted)]
        [NotPersisted]
        public virtual DateTime CreatedDate { get; set; }

        [DisplayName("Last Modified")]
        [Hidden(WhenTo.UntilPersisted)]
        [NotPersisted]
        public virtual DateTime ModifiedDate { get; set; }

        // Even though the database does not record a rowguid for all classes, it does for so many of them 
        // that it is convenient to implement this at the top of the hierarchy.  Classes that don't need it
        // just won't use it, or persist it. Classes that do need it, will still have to override the 
        // implementation to be recognised by EF.
        [Hidden(WhenTo.Always)]
        [NotPersisted]
        public virtual Guid rowguid { get; set; }

        [Hidden(WhenTo.Always)]
        public virtual void Seed()
        {
            rowguid = Guid.NewGuid();
            CreatedDate = DateTime.Now;
            ModifiedDate = CreatedDate;
        }
    }
}
        }

**Comment by acrigney** [comment]
Awesome yes I had forgotten to change them!
I put the ConstantStringHasher in and I am also trying to denormalise the model.

container.RegisterType<IStringHasher, ConstantStringHasher>(new PerResolveLifetimeManager());
Maybe this would fix it?

**Comment by acrigney** [comment]
As the choces dropdown was not working when I deployed to IIS I changed all of the code and choices to use strings and it works. Except as soon as I have another user on the system I get
These transient integrity errors.
Can I turn that off?
Here is an example from the log file.
ERROR 7 NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity failed for: TEOID#178 bad values: UserName:WorkItemId:0Creator:WorkItemState:PendingADUserName:AuthorisingOfficerEmail:CreatedDate:11/09/2018 12:00:00 AMIsWorkItemApproved:FalseModifiedDate:11/09/2018 12:00:00 AMUserName:WorkItemOfficerEmail:rowguid:31839299-636f-4519-a129-6b3fce03705d old hash:  new hash

**Comment by acrigney** [comment]
I tried replacing the ConstantStringHasher with this hasher but it didn't work
public class NullStringHasher : IStringHasher
        {
            public NullStringHasher()
            {

            }

            public string GetHash(string toHash)
            {
                return (String.Empty);
            }

        }

**Comment by acrigney** [comment]
Mate this is a show stopper for me, they won't be able to use the system unless multiple users can use the system. Can you give me any more tips?

I tried to build the code from your instructions so I can debug directly but I had to reinstall the packages and the get-help nakedobjects etc is not working for me.

Update-Package  –reinstall


Attempting to gather dependency information for multiple packages with respect to project 'NOFBuild', targeting '.NETFramework,Version=v4.5.2'
Gathering dependency information took 338.48 ms
Attempting to resolve dependencies for multiple packages.
Resolving dependency information took 0 ms
Resolving actions install multiple packages
Retrieving package 'NakedObjects.Ide 9.0.1' from 'nuget.org'.
Removed package 'NakedObjects.Ide 9.0.1' from 'packages.config'
Successfully uninstalled 'NakedObjects.Ide 9.0.1' from NOFBuild
Adding package 'NakedObjects.Ide.9.0.1', which only has tools group, to project 'NOFBuild'
Package 'NakedObjects.Ide.9.0.1' already exists in folder 'D:\Projects\NakedObjectsFramework-master_29082018\NakedObjectsFramework-master\packages'
Added package 'NakedObjects.Ide.9.0.1' to 'packages.config'
Successfully installed 'NakedObjects.Ide 9.0.1' to NOFBuild
Executing nuget actions took 2.57 sec
Time Elapsed: 00:00:03.1633787
PM> get-help nakedobjects
get-help : Get-Help could not find nakedobjects in a help file in this session. To download updated help topics type: "Update-Help". To get help online, search for the help topic in the TechNet library at 
http://go.microsoft.com/fwlink/?LinkID=107116.
At line:1 char:1
+ get-help nakedobjects
+ ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [Get-Help], HelpNotFoundException
    + FullyQualifiedErrorId : HelpNotFound,Microsoft.PowerShell.Commands.GetHelpCommand
 
PM> New-NakedObjectsCleanBuildNoTest
New-NakedObjectsCleanBuildNoTest : The term 'New-NakedObjectsCleanBuildNoTest' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path 
was included, verify that the path is correct and try again.
At line:1 char:1
+ New-NakedObjectsCleanBuildNoTest
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (New-NakedObjectsCleanBuildNoTest:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

**Comment by richardpawson** [comment]
Yes, my tip is:  don't try debugging into the framework.  It should not be necessary, and will only add more confusion.  Try to isolate the problem by programming within the sandbox of the framework as usual.  Create a brand new standalone project (from the NOF9 template) and build up the functionality incrementally, testing it multi-user, until you find the problem.

Or conversely, try eliminating things from the object that is causing the problem until it does work again then build back up.  I would be inclined to start by 'hollowing out' your LifeCycle methods.  Will it persist without those.  Why, for example, are you setting the CreatedDate in Persisted rather than Persisting (I notice it was commented-out from the latter).

When I see comments like 'I tried replacing the ConstantStringHasher' it does make me wonder just how many other workarounds, or unusual coding patterns you might have, that are causing interference.  Transient Integrity Error is not an error that I have encountered, in many years of using the framwork.

**Comment by acrigney** [comment]
Thanks mate yes that was exactly what I was going to do, build up on large
object from a small one.

I have not done any more hacks as you say or workarounds as you say. But
what does a transient integrity error mean, that there has been a
concurrency error? so that somewhere the system is trying to update a stale
object?

**Comment by richardpawson** [comment]
A transient object refers to one that is being persisted _for the first time_ -  whereas concurrency issues would normally arise only when updating an already persisted object.  I assume the error is arising when a user saves a newly created object  -  but please confirm this.  However, if you are doing other things in the lifecycle methods then it is possible that the action of saving a transient object might be causing a conflict in a persisted object, perhaps.  

Also, what do you mean when you say that the problem only occurs when you have more than one user?  Do you mean that you can have more than one user working on it separately, but not if their actions overlap?  Are you using concurrency checking (I didn't see any concurrency check attribute, for example).

**Comment by acrigney** [comment]
Yes the error occurs when a user tries to create any new object. Its like
the system only allows 1 user. But I just tested it now and I was using
different browsers!
It works when I use only instances of the 1 browser type but not when I use
say edge and chrome.
Yes I am creating new objects in the lifecycle methods.
I get the usual concurrency errors when a user tries to update a dirty
object, that has always worked fine.
I have a concurrency check Timestamp on my BaseDomainObject.
Looks like we are getting there thanks heaps!!

**Comment by acrigney** [comment]
By the way the transient error did not actually show me the new hash value,
I put in the missing
ERROR 32 NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity
failed for: TEOID#1278 bad values:
UserName:WorkItemId:0CourtType:UnspecifiedCreator:WorkItemState:PendingADUserName:AuthorisingOfficerEmail:CreatedDate:12/09/2018
12:00:00 AMIsWorkItemApproved:FalseModifiedDate:12/09/2018 12:00:00
AMPractitionerContext:Timestamp:UserName:WorkItemOfficerEmail:rowguid:1ff641c5-ab49-47a8-bf1f-cc40182cf203
old hash:  new hash* ????*

the code has
if (transientHash != digest) {
                Log.Error($"Transient Integrity failed for:
{nakedObject.Id} bad values: {rawValue} old hash: {digest} new hash
{transientHash}");
                throw new BadRequestNOSException("Values provided may not
be persisted as an object (ensure any derived properties are annotated
NotPersisted");

**Comment by richardpawson** [comment]
Please make an effort to be clearer.  Are you saying that it works fine with:

- Multiple users who all use Chrome
- Multiple users who all use Edge

But not when two users are using the system, one in Chrome and one in Edge -  even though the two users are creating different objects?  And if so, is this only if their transactions are overlapping, or even if they use it sequentially i.e. one does a transaction, then the other does one?

Please do some thorough and rigorous testing and articulate a clear statement of what specific circumstances give rise to the problem before posting again. I cannot respond to repeated posts each revealing some new fragment of information.

**Comment by acrigney** [comment]
Sorry I will test more tomorrow but yes this is what I have found so far.

Its working for me with multiple instances of Chrome and with multiple
instances of Edge.
But its not working for me when I use Edge and Chrome at the same time and
I am not testing working on the same objects, saving any object in another
browser does not work.

I will get our devs and testers to test more tomorrow.

Thanks heaps for your support!

**Comment by richardpawson** [comment]
I will just add this.  The transient integrity checking is intended to guard against the client having altered something that it shouldn't.  A transient object may contain fields that cannot be edited, or maybe even hidden, and if those come back different to the ones that were sent then this could indicate attempted security hack -  hence the comparison of hashes. 

One thing you could do is inspect the RESTful Objects (JSON) representations going back and forth (e.g. using Fiddler)   - for the case where it works and where it doesn't work. Just to the same action in both cases, keeping it as simple as possible.  The uneditable and user-hidden fields will still be there in plain form (unless they are explicitly encrypted). 

Look out for any field contents in the JSON that have changed between the out-bound and in-bound versions of the transient object, for fields that the user didn't enter.  If there is an issue between browsers then there must be something different coming back from the two different browsers for the server to detect an integrity issue.  Though why it would work OK when using both browsers independently is certainly strange.

**Comment by acrigney** [comment]
Awesome mate yes of course about the integrity!
But thats a great idea to check the json I will do it tomorrow.
Thanks  again!

**Comment by acrigney** [comment]
I must have fixed something now as now am not getting the integrity errors any more, and I created a simple object that worked fine locally using Edge but not with Chrome on IIS. I used fiddler to detect the differences between the objects sent went its working and when it failed to save and the only difference was the browser type.
The business is probably happy to use Edge, they will just have to use the Edge browser for links to the app from the intranet site.
So I have the following situation.
                        Localhost  IIS 
Chrome           Yes             No
IE                      No            No
Edge                 Yes           Yes
So we are getting there! 
Edge will do I think thanks again!
